### ISSUE TITLE
rename effectively removes server side encryption!

### ISSUE THREAD

**Issue by lutnos** [Rhinofly/play-s3#47]
Hi

I see that rename is implemented as copy then delete.

But the copy made does not mirror any server side encryption of the source and there is now way to specify encryption.

It would be _very_ _desirable_ to be able to specify the encryption of the destination.

It appears that this could be done by making use of Server-Side Encryption Specific Request Headers such as x-amz-server-side​-encryption in fly.play.s3.S3.putCopy .

http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectCOPY.html

But also it needs to be made clear that the present innocuously named rename can be so dangerous.

**Comment by EECOLOR** [comment]
Thank you for this report.

The part of the documentation that we have missed:
> If copied, the metadata, except for the version ID, remains unchanged. In addition, the server-side-encryption, storage-class, and website-redirect-location metadata from the source is not copied.

For now, as a workaround instead of using `putCopy`, a `GET` and `PUT` can be used.

It seems we should automatically add the headers from the bucket the same way we do with `put`. We have to see if this gives any conflicts and we need to add a test to make sure that, once fixed, will not be a problem again.

As soon as I have time to work on this I will pick it up. If you have time yourself, you could try to create test with a pull request and experiment with the solution.

**Comment by lutnos** [comment]
Thanks, yes

Your "get and put" workaround would work, of course, but for us the cost of streaming the content could be a problem.

In our case, we do not actually need to actively copy the encryption header, we know what the header is & should remain - so we can specify it at the rename time.  So I will look to to workaround for our problem by using a signer to add the encryption header.

I can see how your solution can work, though there will be a few cases to cover, to include customer-provided encryption keys and kms.  I'm not yet familiar with all of those cases.  Also it will add an extra latency, of course.

If you intend to support encryption, then you might also want to pay attention to the fact that the aws putCopy seems to me to be the one way of applying encryption to an existing object, or to change they key etc. .  If you were to add provision for specifying headers on your low-level putCopy, then that could be used by those needing to manipulate encrypted objects (including rename) - and that may be convenient for you if augmenting your (growing) rename implementation turns out to be problematic.

All the best

**Comment by EECOLOR** [comment]
In the 6.0.0 branch, the implementation of `putCopy` is a bit easier: https://github.com/Rhinofly/play-s3/blob/6.0.0/src/main/scala/fly/play/s3/S3.scala#L169

val source = "/" + sourceBucketName + "/" + sourcePath

```scala
    client
      .resourceRequest(destinationBucketName, destinationPath)
      .withHeaders("X-Amz-acl" -> acl.value)
      .withHeaders("X-Amz-copy-source" -> source)
      .put("")
```

Using this code outside of the S3 class with extra headers should be trivial.

The `6.0.0` branch is released with version `6.0.0-RC1`.

**Comment by EECOLOR** [comment]
@lutnos Would it be enough for your usecase to augment [the signature of `rename`](https://github.com/Rhinofly/play-s3/blob/6.0.0/src/main/scala/fly/play/s3/Bucket.scala#L132) into:

```
  def rename(
    sourceItemName: String,
    destinationItemName: String,
    acl: ACL = PUBLIC_READ,
    additionalHeaders: Map[String, String]
  ): Future[Unit]
```
